catfs/mio/compress/algorithm.go

```diff
@@ -50,74 +55,135 @@ func (at AlgorithmType) String() string {
 
 // Algorithm is the common interface for all supported algorithms.
 type Algorithm interface {
-	Encode([]byte) ([]byte, error)
-	Decode([]byte) ([]byte, error)
+	Encode(dst, src []byte) ([]byte, error)
+	Decode(dst, src []byte) ([]byte, error)
+	MaxEncodeBufferSize() int
 }
 
 type snappyAlgo struct{}
-type lz4Algo struct{}
+
+type lz4Algo struct {
+	compressor *lz4.Compressor
+}
+
+type zstdAlgo struct{}
+
+var (
+	zstdWriter *zstd.Encoder
+	zstdReader *zstd.Decoder
+)
+
+func init() {
+	var err error
+
+	// NOTE: zstd package allows us to use the same writer and reader
+	//       stateless if we just use block encoding/decoding.
+	//       This saves us some extra allocations.
+
+	// TODO: configure compression level?
+	zstdWriter, err = zstd.NewWriter(
+		nil,
+		zstd.WithEncoderLevel(zstd.SpeedDefault),
+	)
+
+	if err != nil {
+		// configuring the writer wrong is a programmer error.
+		panic(err)
+	}
+
+	// NOTE: reader should set max memory bound with WithDecoderMaxMemory.
+	//       we can deduce it from maxChunkSize and protect against
+	//       malicious inputs.
+	zstdReader, err = zstd.NewReader(
+		nil,
+		zstd.WithDecoderMaxMemory(32*maxChunkSize),
+	)
+	if err != nil {
+		// configuring the reader wrong is a programmer error.
+		panic(err)
+	}
+}
 
 var (
 	// AlgoMap is a map of available algorithms.
-	AlgoMap = map[AlgorithmType]Algorithm{
-		AlgoSnappy: snappyAlgo{},
-		AlgoLZ4:    lz4Algo{},
+	algoMap = map[AlgorithmType]func() Algorithm{
+		AlgoSnappy: func() Algorithm {
+			return snappyAlgo{}
+		},
+		AlgoLZ4: func() Algorithm {
+			// TODO: we could configure compression level here.
+			return &lz4Algo{
+				compressor: &lz4.Compressor{},
+			}
+		},
+		AlgoZstd: func() Algorithm {
+			return zstdAlgo{}
+		},
 	}
 
-	// TODO: still needded?
 	algoToString = map[AlgorithmType]string{
 		AlgoSnappy: "snappy",
 		AlgoLZ4:    "lz4",
-	}
-
-	// TODO: still needed?
-	stringToAlgo = map[string]AlgorithmType{
-		"snappy": AlgoSnappy,
-		"lz4":    AlgoLZ4,
+		AlgoZstd:   "zstd",
 	}
 )
 
 // AlgoSnappy
-func (a snappyAlgo) Encode(src []byte) ([]byte, error) {
-	return snappy.Encode(nil, src), nil
+func (a snappyAlgo) Encode(dst, src []byte) ([]byte, error) {
```

Comment:
It is purely to satisfy my curiosity: but don't we already have `dst` with all the data. What is the need to return the byte range? It clearly crucial, but I do not see why.

Reply:
The ``dst`` is the complete buffer, returned is a sub slice of ``dst`` where ``len()`` indicates the length of the compressed (or decompressed) data. This is always true though: ``len(returned) <= len(dst)``.